test/utils/readRowsImpl.ts

```diff
@@ -0,0 +1,266 @@
+        chunk.familyName = {
+          value: 'family',
+        };
+        chunk.qualifier = {
+          value: Buffer.from('qualifier').toString('base64'),
+        };
+      }
+      const thisChunkSize =
+        remainingBytes >= chunkSize ? chunkSize : remainingBytes;
+      remainingBytes -= thisChunkSize;
+      const value = Buffer.from('a'.repeat(remainingBytes)).toString('base64');
+      chunk.value = value;
+      if (remainingBytes === 0) {
+        debugLog(`  setting commit flag for rowKey ${key}`);
+        chunk.commitRow = true;
+      }
+      chunks.push(chunk);
+      ++chunkCounter;
+    }
+  }
+  debugLog(`generated ${chunks.length} chunks between ${keyFrom} and ${keyTo}`);
+  return chunks;
+}
+
+export function readRowsImpl(
+  keyFrom: number,
+  keyTo: number,
+  errorAfterChunkNo?: number
+) {
+  return async (
+    stream: ServerWritableStream<
+      protos.google.bigtable.v2.IReadRowsRequest,
+      protos.google.bigtable.v2.IReadRowsResponse
+    >
+  ) => {
+    // pretty-printing important parts of the request.
+    // doing it field by field because we want to apply .toString() to all key fields
+    debugLog('received request: {');
+    debugLog(`  tableName: "${stream.request.tableName}",`);
+    if (stream.request.rows) {
+      debugLog('  rows: {');
+      if (stream.request.rows.rowKeys) {
+        debugLog('    rowKeys: [');
+        for (const key of stream.request.rows.rowKeys) {
+          debugLog(`      "${key.toString()}",`);
+        }
+        debugLog('    ],');
+      }
+      if (stream.request.rows.rowRanges) {
+        debugLog('    rowRanges: [');
+        for (const range of stream.request.rows.rowRanges) {
+          debugLog('      {');
+          if (range.startKeyOpen) {
+            debugLog(
+              `        startKeyOpen: "${range.startKeyOpen.toString()}",`
+            );
+          }
+          if (range.startKeyClosed) {
+            debugLog(
+              `        startKeyClosed: "${range.startKeyClosed.toString()}",`
+            );
+          }
+          if (range.endKeyOpen) {
+            debugLog(`        endKeyOpen: "${range.endKeyOpen.toString()}",`);
+          }
+          if (range.endKeyClosed) {
+            debugLog(
+              `        endKeyClosed: "${range.endKeyClosed.toString()}",`
+            );
+          }
+          debugLog('      },');
+        }
+        debugLog('    ],');
+      }
+      debugLog('  },');
+    }
+    debugLog('}');
+
+    let stopWaiting: () => void = () => {};
+    let cancelled = false;
+
+    stream.on('cancelled', () => {
+      debugLog('gRPC server received cancel()');
+      cancelled = true;
+      stopWaiting();
+      stream.emit('error', new Error('Cancelled'));
+    });
+
+    let chunksSent = 0;
+    const chunks = generateChunks(keyFrom, keyTo, stream);
+    let lastScannedRowKey: string | undefined;
+    let firstN: protos.google.bigtable.v2.ReadRowsResponse.ICellChunk[] = [];
```

Comment:
Should we rename this variable to `downstreamChunks` or something?

Reply:
Yeah I was struggling with finding a good name, since `chunks` is already used :)  `downstreamChunks` or `responseChunks` maybe. I'll rename.

Reply:
I renamed to `currentResponseChunks`, does it make more sense now?